stdlib/public/core/String.swift

```diff
@@ -493,6 +493,112 @@ extension String {
     self = String._fromNonContiguousUnsafeBitcastUTF8Repairing(codeUnits).0
   }
 
+  /// Creates a new string by copying and validating the sequence of
+  /// code units passed in, according to the specified encoding.
+  ///
+  /// This initializer does not try to repair ill-formed code unit sequences.
+  /// If any are found, the result of the initializer is `nil`.
+  ///
+  /// The following example calls this initializer with the contents of two
+  /// different arrays---first with a well-formed UTF-8 code unit sequence and
+  /// then with an ill-formed UTF-16 code unit sequence.
+  ///
+  ///     let validUTF8: [UInt8] = [67, 97, 0, 102, 195, 169]
+  ///     let valid = String(validating: validUTF8, as: UTF8.self)
+  ///     print(valid)
+  ///     // Prints "Optional("Café")"
+  ///
+  ///     let invalidUTF16: [UInt16] = [0x41, 0x42, 0xd801]
+  ///     let invalid = String(validating: invalidUTF16, as: UTF16.self)
+  ///     print(invalid)
+  ///     // Prints "nil"
+  ///
+  /// - Parameters:
+  ///   - codeUnits: A sequence of code units that encode a `String`
+  ///   - encoding: A conformer to `Unicode.Encoding` to be used
+  ///               to decode `codeUnits`.
+  @inlinable
+  @available(SwiftStdlib 5.10, *)
+  public init?<Encoding: Unicode.Encoding>(
+    validating codeUnits: some Sequence<Encoding.CodeUnit>,
+    as encoding: Encoding.Type
+  ) {
+    let newString: String?? = codeUnits.withContiguousStorageIfAvailable {
+      String._validate($0, as: Encoding.self)
+    }
+    if let newString {
+      guard let newString else { return nil }
+      self = newString
+      return
+    }
+
+    // slow-path
+    var transcoded: [UTF8.CodeUnit] = []
```

Comment:
Ohhh we can't do a dry run because this takes a Sequence, not a Collection 😔.

But it's pointless - we're not going to do this in a single pass; we're just going to copy internally anyway. This API should've taken a Collection, IMO.

We can [dispatch](https://swift.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:swift,selection:(endColumn:46,endLineNumber:16,positionColumn:46,positionLineNumber:16,selectionStartColumn:46,selectionStartLineNumber:16,startColumn:46,startLineNumber:16),source:'func+test()+%7B%0A++++doSomething(%5B1,+2,++3%5D)%0A%7D%0A%0Afunc+doSomething(_+x:+some+Sequence%3CUInt8%3E)+%7B%0A++++if+let+collection+%3D+x+as%3F+any+Collection%3CUInt8%3E+%7B%0A++++++++useCollection(collection)%0A++++%7D%0A++++useSequence(x)%0A%7D%0A%0A@_silgen_name(%22useSequence%22)%0Afunc+useSequence(_:+some+Sequence%3CUInt8%3E)%0A%0A@_silgen_name(%22useCollection%22)%0Afunc+useCollection(_:+some+Collection%3CUInt8%3E)'),l:'5',n:'0',o:'Swift+source+%231',t:'0')),k:39.1644908616188,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:swift58,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:swift,libs:!(),options:'-O',overrides:!(),selection:(endColumn:32,endLineNumber:60,positionColumn:32,positionLineNumber:60,selectionStartColumn:32,selectionStartLineNumber:60,startColumn:32,startLineNumber:60),source:1),l:'5',n:'0',o:'+x86-64+swiftc+5.8+(Editor+%231)',t:'0')),k:60.835509138381205,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) based on whether or not the type is a collection, but the optimiser does a poor job specialising it (#62264). I'd still suggest we do that, though, and let the optimiser catch up. The existential wrapping and unwrapping overhead is likely less than allocating and copying everything to an array, and one day the compiler will just eliminate that overhead entirely.

Reply:
There is quite a bit of improvement to be done to the internal transcoding machinery, and this slowest path will then be adapted to use that. It would be better to allocate a string buffer directly and resize that, but it is not possible at the moment.

Reply:
My understanding is that it is fairly common in C APIs (e.g. [`wcsrtombs`](https://pubs.opengroup.org/onlinepubs/9699919799/), [`WideCharToMultiByte`](https://learn.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-widechartomultibyte), etc) to perform a dry run by specifying the output buffer as `NULL`, getting the length, and converting in to an appropriately-sized buffer. That's why I suggested it.

Resizing is less than ideal because it's quadratic. Array mitigates this somewhat with an over-allocation strategy that scales geometrically, but that is also wasteful, and I don't think String employs the same strategy (?).

For Sequence, where we can only make one pass, we unfortunately _have_ to transcode in to some kind of resizing buffer. For Collection, we can just do two passes, guaranteeing no resizing and lower memory water-mark.